### PR TITLE
Allow using tagged final functor in parallel_reduce

### LINKED ISSUE #3651: Support work tag overload final() function in parallel dispatch functor
Currently functor used for parallel dispatch can't have a work-tagged final() function, in contrast to the init() and join(), which can be work-tagged

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes #3651.

### CODE REVIEW

File: core/src/impl/Kokkos_FunctorAdapter.hpp [hunk 1]
```diff
@@ -2091,13 +2091,33 @@ struct FunctorFinal<FunctorType, ArgTag,
                     decltype(
                         FunctorFinalFunction<FunctorType, ArgTag>::enable_if(
                             &FunctorType::final))> {
-  KOKKOS_FORCEINLINE_FUNCTION static void final(const FunctorType& f, void* p) {
+  template <typename Dummy = ArgTag>
+  KOKKOS_FORCEINLINE_FUNCTION static std::enable_if_t<
+      std::is_same<Dummy, void>::value>
```
Comment (dalg24) [inline core/src/impl/Kokkos_FunctorAdapter.hpp:2096]: This works but departs from what is done above for `FunctorValueInit` and `FunctorValueJoin` where two separate specializations are provided.

File: core/unit_test/TestReduce.hpp [hunk 2]
```diff
@@ -86,12 +88,27 @@ class ReduceFunctor {
     dst.value[2] += src.value[2];
   }
 
+  KOKKOS_INLINE_FUNCTION
+  void join(const ReducerTag, volatile value_type& dst,
+            const volatile value_type& src) const {
```
Comment (dalg24) [inline core/unit_test/TestReduce.hpp:93]: Probably should either do something different from the overload that does not take a work tag or define another class.

File: core/src/impl/Kokkos_FunctorAdapter.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/impl/Kokkos_FunctorAdapter.hpp:2096]: Yes, this seems to be less code so I went with this approach.

Comment (masterleinad) [reply core/src/impl/Kokkos_FunctorAdapter.hpp:2096]: If you insist, I can change, of course.

File: core/unit_test/TestReduce.hpp (hunk 2 above)
Comment (masterleinad) [reply core/unit_test/TestReduce.hpp]: I opted for not deriving from `ReduceFunctor`.

**Comment by masterleinad** [comment]
Ready from my side. Needs more reviews.

**Review by crtrott — APPROVED** [review]

File: core/src/impl/Kokkos_FunctorAdapter.hpp (hunk 1 above)
Comment (dalg24) [reply core/src/impl/Kokkos_FunctorAdapter.hpp:2096]: I would prefer consistency within that file but I will not block if I am the only one feeling that way.

**Review by dalg24 — APPROVED** [review]

**Pull request merged**